lua/bson_lua_wrapper.c

```diff
@@ -0,0 +1,372 @@
+#include <stdlib.h>
+#include <string.h>
+ 
+#include <lua5.2/lua.h>
+#include <lua5.2/lauxlib.h>
+ 
+#include <bson_object.h>
+
+static int table_to_bson_object(lua_State *L, BsonObject *obj, char *errorMessage);
+static int table_to_bson_array(lua_State *L, BsonArray *arr, char *errorMessage);
+
+static int bson_object_to_table(lua_State *L, BsonObject *obj, char *errorMessage);
+static int bson_element_to_table(lua_State *L, BsonElement *element, char *errorMessage);
+static int bson_array_to_table(lua_State *L, BsonArray *arr, char *errorMessage);
+
+static int lbson_object_to_bytes(lua_State *L) {
+    BsonObject bson;
+    bson_object_initialize_default(&bson);
+    if (!lua_istable(L, 1)) {
+        return luaL_argerror(L, 1, "Arg 1 must be a table");
+    }
+
+    char errorMessage[255];
+    int success = table_to_bson_object(L, &bson, errorMessage);
+    if (success != 0) {
+        bson_object_deinitialize(&bson);
+        return luaL_argerror(L, 1, errorMessage);
+    }
+
+    uint8_t *bytes = bson_object_to_bytes(&bson);
+    if (bytes == NULL) {
+        bson_object_deinitialize(&bson);
+        return luaL_argerror(L, 1, "Map could not be parsed into bson");
+    }
+
+    size_t obj_size = bson_object_size(&bson);
+    bson_object_deinitialize(&bson);
+
+    char *byte_string = malloc(obj_size * sizeof(char));
+    size_t index = 0;
+    for (index = 0; index < obj_size; index++) {
+        byte_string[index] = bytes[index] & 0xFF;
+    }
+    free(bytes);
+
+    lua_pushlstring(L, (const char *)byte_string, obj_size);
+    free(byte_string);
+
+    return 1;
+}
+
+static int table_to_bson_object(lua_State *L, BsonObject *obj, char *errorMessage) {
+    lua_pushnil(L); // Stack: [bson_table, nil]
+    while (lua_next(L, -2) != 0) { //Stack: [bson_table, key, {type, value}]
+        if (!lua_istable(L, -1)) {
+            snprintf(errorMessage, 255, "Table is not in BSON format (key: {type, value})");
+            return -1;
+        }
+        const char *key = luaL_checkstring(L, -2);
+
+        lua_getfield(L, -1, "type"); //Stack: [bson_table, key, {type, value}, type]
+
+        lua_Integer type = luaL_checkinteger(L, -1);
+        lua_pop(L, 1); //Stack: [bson_table, key, {type, value}]
+
+        lua_getfield(L, -1, "value"); //Stack: [bson_table, key, {type, value}, value]
+        switch ((element_type)type) {
```

Comment:
@jacobkeeler Please add handling for NULL type of BSON element. 
NULL type is present in  bson_util.h and Lua Supported BSON Types enumeration according Constructed Payloads Testing proposal.

Reply:
Considering that there is no public `bson_object_put` or `bson_object_put_null` functions, I would consider the support of the `NULL` type to be a misleading mistake. I will add the `unused` comment to this type.

Reply:
@jacobkeeler Ok, but we had a problems with unsupported NULL in json4lua couple times during testing of different RPCs. And mentioned support was added. That is why I made my proposition.

Reply:
Agreed, and if needed we can support it in the future with a `bson_object_put_null` method, but for now it is actually unused.